### ISSUE TITLE
Workspace Hierarchy Request 

### ISSUE THREAD

**Issue by gorkem** [Microsoft/language-server-protocol#136]
It is important for object oriented languages to retrieve the inheritance information. I would like to discuss a new request for retrieving hierarchy of a given Symbol.

**Comment by felixfbecker** [comment]
Some food for thought: Many languages have a fully qualified structural element name for every symbol. Using such a unique identifier, you could represent the tree like this:
```ts
interface SymbolInformation {
  // other properties...
  /** fully qualified structural element name (must be unique) */
  fqsen?: string;
  /** an array of symbol FQSENs that this symbol is a child of in a class hierachy */
  extends?: string[];
}
```
You would then get the hierarchy from `workspace/symbol` or `textDocument/documentSymbol` or any method that returns `SymbolInformation`

**Comment by dbaeumer** [comment]
I agree that a hierarchy request is desirable.  We could add this without breaking clients since clients that can't render this will not request it.

**Comment by tsmaeder** [comment]
@felixfbecker if we go by the example of Java, fqn's are not unique, neither at development time (check out multiple versions of the same class) nor at runtime (load different versions of the same class twice, in different class loaders).
I belive we need something like ```extends: SymbolInformation [];```
Also, what about subtype hierarchy?

**Comment by dbaeumer** [comment]
I worked on contributing proposed protocol more easily. A document describing this is here:  https://github.com/Microsoft/language-server-protocol/blob/master/contributing.md

May be someone wants to look into this.

**Comment by gorkem** [comment]
Initial proposal posted https://github.com/Microsoft/vscode-languageserver-node/pull/346

**Comment by dbaeumer** [comment]
I added my comments in Microsoft/vscode-languageserver-node#346 since it contains a concrete proposal.

**Comment by KamasamaK** [comment]
The current proposal for this feature is at #1231. Feel free to provide any feedback you might have.

**Comment by dbaeumer** [comment]
Added to 3.17.
